### PR TITLE
PSR Config

### PR DESCRIPTION
**Pull request by sandrokeil**
Please take a look at the [config-meta.md](https://github.com/sandrokeil/fig-standards/blob/config/proposed/config-meta.md) and [config.md](https://github.com/sandrokeil/fig-standards/blob/config/proposed/config.md) for details. This is a draft, so if you have any suggestions, please leave a comment.

### CODE REVIEW

**Comment by sandrokeil** [comment]
Replacing *component* with *package* makes more sense, because it's more like Java.

Any thoughts on this?

**Comment by sandrokeil** [comment]
I've added two new interfaces `HasDefaultOptions` and `HasOptionalOptions`. They are useful for auto discovery of factories and for default values of options.

**Comment by sandrokeil** [comment]
I've renamed the interfaces for better understanding and reduced the complexity by removing `ÒbtainsOptions` and `HasDefaultOptions` interface.

The `PSR\Config\RequiresConfig` is the main interface and `PSR\Config\RequiresContainerId`, `PSR\Config\RequiresMandatoryOptions` and `PSR\Config\ProvidesDefaultOptions` are optional.

There is a [quick start guide](https://github.com/sandrokeil/interop-config/blob/master/doc/book/quick-start.md) which also explains the need.

File: proposed/config-meta.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+## 1. Introduction
+This document describes the process and discussions that led to the Config PSR. It's goal is to explain the reasons 
+behind each decision.
+
+The word `Container` in this document is to be interpreted as the `ContainerInterface` of the PSR Container proposal.
+
+## 2. Why bother?
+The configuration of instances via factories is not uniform, uses different ways to retrieve and check the needed 
+options and it's not clear which configuration options needs the factory to create the instance.
```
Comment (samdark) [inline proposed/config-meta.md]: the factory needs?

File: proposed/config-meta.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+## 3. Scope
+
+### 3.1 Goals
+The goal set by the Config PSR is to standardize how factories uses a configuration to create instances, support for 
+auto discovery of needed configuration, to reduce boilerplate code and to make it more readable and easier to understand. 
+It can also be used to build the name of the *Container* entry.
+
+### 3.2 Non-goals
+* It's not a goal to define a whole project/library configuration structure.
+* It's not a goal to validate the configuration option values.
```
Comment (samdark) [inline proposed/config-meta.md]: Is it a goal to validate structure?

File: proposed/config-meta.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+## 4. History
+Before submitting the Config PSR to the PHP-FIG, the interfaces was first proposed in a project named
```
Comment (samdark) [inline proposed/config-meta.md]: was → were?

File: proposed/config-meta.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+### 3.2 Non-goals
+* It's not a goal to define a whole project/library configuration structure.
+* It's not a goal to validate the configuration option values.
+
+## 4. History
+Before submitting the Config PSR to the PHP-FIG, the interfaces was first proposed in a project named 
+[interop-config](https://github.com/sandrokeil/interop-config).
+
+The goal of the project was to provide a test-bed for implementing the interfaces and implementation and to pave the way
```
Comment (samdark) [inline proposed/config-meta.md]: implementing the interfaces and implementation → implementing the interfaces

File: proposed/config-meta.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+## 4. History
+Before submitting the Config PSR to the PHP-FIG, the interfaces was first proposed in a project named 
+[interop-config](https://github.com/sandrokeil/interop-config).
+
+The goal of the project was to provide a test-bed for implementing the interfaces and implementation and to pave the way 
+for the Config PSR.
+
+## 5. Interfaces
+There are six interfaces and some interfaces extends from another interface to ensure the uniform configuration structure. 
+If you confused about the naming of the interfaces or methods, note that I have used the Domain Driven Design approach.
```
Comment (samdark) [inline proposed/config-meta.md]: you *are* confused

File: proposed/config-meta.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+### 5.1 RequiresConfig
+This is the main interface and describes the default configuration structure. Since we have Composer more and more
```
Comment (samdark) [inline proposed/config-meta.md]: Missing `,` after "Since we have Composer"

File: proposed/config-meta.md [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+for the Config PSR.
+
+## 5. Interfaces
+There are six interfaces and some interfaces extends from another interface to ensure the uniform configuration structure. 
+If you confused about the naming of the interfaces or methods, note that I have used the Domain Driven Design approach.
+
+### 5.1 RequiresConfig
+This is the main interface and describes the default configuration structure. Since we have Composer more and more 
+libraries can be combined to create individual projects. So it is only logical to start the configuration structure 
+with a `vendorName`. A vendor has multiple packages so the next structure depth is `packageName`.
```
Comment (samdark) [inline proposed/config-meta.md]: It's a big hassle to start it with a vendor name when most of the packages are provided by framework and the vendor is the same. It feels like unnecessary complication in this case.

File: proposed/config-meta.md [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+
+The `canRetrieveOptions` method:
+
+* checks if options are available depending on implemented interfaces and checks that the retrieved options are an array 
+or have implemented \ArrayAccess. The `RequiresContainerId` interface is optional but MUST be supported.
+
+### 5.2 RequiresContainerId
+Factories allow to create more than one instance of the same class with another configuration. This can be achieved with 
+the third structure depth `containerId`. Now the structure is fully expanded.
```
Comment (samdark) [inline proposed/config-meta.md]: Couldn't it be achieved by using another package name for the same class instead?

Comment (samdark) [inline proposed/config-meta.md]: It would look much more compact.

File: proposed/config-meta.md (hunk 1 above)
Comment (sandrokeil) [inline proposed/config-meta.md]: Fixed, thanks.

File: proposed/config-meta.md (hunk 2 above)
Comment (sandrokeil) [inline proposed/config-meta.md]: Yes, it validates the configuration structure depending on the implemented interfaces.

File: proposed/config-meta.md (hunk 3 above)
Comment (sandrokeil) [inline proposed/config-meta.md]: Fixed, thanks.

File: proposed/config-meta.md (hunk 4 above)
Comment (sandrokeil) [inline proposed/config-meta.md]: Fixed, thanks.

File: proposed/config-meta.md (hunk 5 above)
Comment (sandrokeil) [inline proposed/config-meta.md]: Yes, sorry. I've fixed it.

File: proposed/config-meta.md (hunk 6 above)
Comment (sandrokeil) [inline proposed/config-meta.md]: Fixed, thanks.

File: proposed/config-meta.md (hunk 7 above)
Comment (sandrokeil) [inline proposed/config-meta.md]: [Zend Framework 2](http://framework.zend.com/blog/zend-framework-2-5-0-released.html) was split into separate repositories and there are many other good PHP libraries. I don't think it is complicated and works also for monolithic frameworks. The config entries should depends on the package name and you find the options for a package easier. Take a look at this example.

```php
return [
    // vendor
    'zend' => [
        // package
        'service-manager' => [
            // options
            'factories' => [ /* ... */],
        ],
        // package
        'navigation' => [
            // container id
            'default' => [ /* ... */],
        ],
        // package
        'view_manager' => [
            // container id
            'template-map-resolver' => [ /* ... */],
        ],
    ],
    // vendor
    'symfony' => [
        // package
        'services' => [
            /* ... */
        ],
    ],
];
```

File: proposed/config-meta.md (hunk 8 above)
Comment (sandrokeil) [inline proposed/config-meta.md]: This is possible, but it makes the config more unreadable and makes no sense. I think the uniform structure is missing in this case, beause you are mixing the package name with an id or something else. It's up to you to use this interface. There is no check, if the vendor/package name matches the current factory. Only the configuration array is checked for vendor/package.

The `RequriesContainerId` interface is only needed for factories which creates multiple instances with different configuration e.g. database connections (read/write). There are many PHP libraries which use this approach.

* [DoctrineORMModule](https://github.com/doctrine/DoctrineORMModule/blob/master/docs/configuration.md)
* [Prooph Service Bus](https://github.com/prooph/service-bus/blob/develop/config/prooph_service_bus.config.php)
* Zend Framework 2 and other frameworks in some cases

File: proposed/config-meta.md (hunk 2 above)
Comment (samdark) [inline proposed/config-meta.md]: It's a good to mention it.

**Comment by michaelcullum** [comment]
We discuss new proposals for PSRs on the mailing list first normally. If you'd like to continue with this, please suggest it on there where we can discuss it going to a potential entrance vote.

https://groups.google.com/forum/#!forum/php-fig

**Pull request closed without merging**